analysis/classification/utils/db_functions.py

```diff
@@ -226,20 +235,56 @@ def assign_sample_data_to_label(token='admin'):
 
 
 def get_datapoint_to_label(labeler):
-    try:
-        bq_client.get_table(to_label_table_id)
-    except (NotFound, Forbidden):
-        print("NO DATA TO LABEL")
+    if "data_to_label" in st.session_state and len(st.session_state.data_to_label) >= 5:
+        # if we have data in the cache, no need to fetch more data
+        pass
+    else:
+        if "fetch_job" in st.session_state and st.session_state.fetch_job is not None:
+            # see if job is done
+            if st.session_state.fetch_job.done():
+                # load new data from completed job
+                new_data = st.session_state.fetch_job.result().to_dataframe()
+                if "data_to_label" not in st.session_state:
+                    st.session_state['data_to_label'] = new_data
+                else:
+                    st.session_state.data_to_label = pd.concat([st.session_state.data_to_label, new_data])
+                st.session_state.fetch_job = None
+            else:
+                # we may or may not need to wait for job to finish
+                pass
+        else:
+            # need to fetch data
+            try:
+                st.session_state.bq_client.get_table(to_label_table_id)
+            except (NotFound, Forbidden):
+                st.error("NO DATA AVAILABLE TO LABEL")
+                return None
+            fetch_job = None
+            if table_exists(labelled_table_id):
+                fetch_job = st.session_state.bq_client.query(
+                    f"SELECT * FROM {to_label_table_id} a LEFT JOIN {labelled_table_id} using(id_a, id_b)  where a.labeler='{labeler}' AND label IS NULL LIMIT 20")
+            else:
+                fetch_job =  st.session_state.bq_client.query(
+                    f"SELECT * FROM {to_label_table_id}  where labeler='{labeler}' LIMIT 20")
```

Comment:
So, we're assigning 20 data points to each RA at once?

Reply:
Yeah, basically we're loading a cache of 20 points so we don't need to hit the db for every point.  If they don't all get labelled, that's fine, they'll still be eligible to be selected in future sessions.